File: leetcode-python-go/790. 多米诺和托米诺平铺/main.py
from typing import List

n = 3

def numTilings(n: int) -> int:
    MOD = 10 ** 9 + 7
    dp = [[0] * 4 for _ in range(n + 1)]
    dp[0][3] = 1
    for i in range(1, n + 1):
        dp[i][0] = dp[i - 1][3]
        dp[i][1] = (dp[i - 1][0] + dp[i - 1][2]) % MOD
        dp[i][2] = (dp[i - 1][0] + dp[i - 1][1]) % MOD
        dp[i][3] = (((dp[i - 1][0] + dp[i - 1][1]) % MOD + dp[i - 1][2]) % MOD + dp[i - 1][3]) % MOD
    return dp[n][3]


print(numTilings(n))

File: leetcode-python-go/790. 多米诺和托米诺平铺/main.go
package main

import (
	"fmt"
)

func numTilings(n int) int {
	const mod int = 1e9 + 7
	dp := make([][4]int, n+1)
	dp[0][3] = 1
	for i := 1; i <= n; i++ {
		dp[i][0] = dp[i-1][3]
		dp[i][1] = (dp[i-1][0] + dp[i-1][2]) % mod
		dp[i][2] = (dp[i-1][0] + dp[i-1][1]) % mod
		dp[i][3] = (((dp[i-1][0]+dp[i-1][1])%mod+dp[i-1][2])%mod + dp[i-1][3]) % mod
	}
	return dp[n][3]
}

func main() {
	n := 3
	results := numTilings(n)
	fmt.Println(results)
}
